pkg/registry/cluster/strategy.go

```diff
@@ -0,0 +1,122 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cluster
+
+import (
+	"fmt"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/apis/controlplane"
+	"k8s.io/kubernetes/pkg/apis/controlplane/validation"
+	"k8s.io/kubernetes/pkg/fields"
+	"k8s.io/kubernetes/pkg/labels"
+	"k8s.io/kubernetes/pkg/registry/generic"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/util/validation/field"
+)
+
+type clusterStrategy struct {
+	runtime.ObjectTyper
+	api.NameGenerator
+}
+
+var Strategy = clusterStrategy{api.Scheme, api.SimpleNameGenerator}
+
+func (clusterStrategy) NamespaceScoped() bool {
+	return false
+}
+
+func ClusterToSelectableFields(cluster *controlplane.Cluster) fields.Set {
+	objectMetaFieldsSet := generic.ObjectMetaFieldsSet(cluster.ObjectMeta, false)
+	specificFieldsSet := fields.Set{
+		"status.phase": string(cluster.Status.Phase),
+	}
+	return generic.MergeFieldsSets(objectMetaFieldsSet, specificFieldsSet)
+}
+
+func MatchCluster(label labels.Selector, field fields.Selector) generic.Matcher {
+	return &generic.SelectionPredicate{
+		Label: label,
+		Field: field,
+		GetAttrs: func(obj runtime.Object) (labels.Set, fields.Set, error) {
+			cluster, ok := obj.(*controlplane.Cluster)
+			if !ok {
+				return nil, nil, fmt.Errorf("given object is not a cluster.")
+			}
+			return labels.Set(cluster.ObjectMeta.Labels), ClusterToSelectableFields(cluster), nil
+		},
+	}
+}
+
+// PrepareForCreate clears fields that are not allowed to be set by end users on creation.
+func (clusterStrategy) PrepareForCreate(obj runtime.Object) {
+	cluster := obj.(*controlplane.Cluster)
+	cluster.Status = controlplane.ClusterStatus{
+		Phase: controlplane.ClusterPending,
```

Comment:
We can make this the default Phase value?
I was thinking that we can handle this the same way as we handle NodePending, but it seems like we neither set a default NodePhase, nor modify it in CreateStrategy.
Setting default Phase value seems better to me